app/components/modal_component.html.erb

```diff
@@ -0,0 +1,13 @@
+<%= turbo_frame_tag "modal" do %>
+  <%= tag.div id:'modal-overlay',
+              class: 'fixed z-20 w-screen h-screen bg-gray-100/75',
+              data: { controller: 'modal',
+                      modal_target: 'modal',
+                      action: 'turbo:submit-end->modal#submitEnd keydown.esc@window->modal#closeWithEscape mousedown@window->modal#closeWithBackgroundClick'} do %>
```

Comment:
🤮

Reply:
🤣 I know, it looks awful, and the connection with the actual controller actions is so tenuous. 

I was thinking of writing a helper module for stuff like this to make it more readable. It would be really cool if we could make something like Phoenix's [verified_routes](https://www.germanvelasco.com/blog/phoenix-1-7-verified-routes). That way we wouldn't need to have magic strings. We could round up all the controller actions in the JS at compile time and set a global verified_stimulus_actions map. The tough part would be somehow generating all the valid event triggers to go with them... maybe we wouldn't have to though. Ending up with something like this:
```ruby
$verified_stimulus_actions = { 'modal': ['closeEnd', 'submitEnd'] }
```
... then have a macro/resolver of some kind check if the action exists and fail more visibly if it does not:
```ruby
module StimulusHelper
  def verify_action (action)
    check_for_action($verified_stimulus_actions)
  end
end
```
...Something like that anyway. 

In the meantime, I was thinking something like this just for readability:

```ruby
module StimulusHelper
  def action_list(actions)
    actions.join(' ')
  end
end

...
### modal_component.html.erb
<tag.div id:'modal-overlay;
  data: {
    controller: 'modal',
    action: StimulusHelper.action_list(["turbo:submit-end->modal#submitEnd", "keydown.esc@window->modal#closeWithEscape"])
  } do %>
  <%= content %>
  <% end %>
```

Reply:
Gonna make an issue for this actually. Don't want to hold up this PR any longer but the more I think about it the better an idea this seems. Maybe even a gem worth publishing, if I can't find one that does something similar 🤔